crates/iota-core/src/authority/shared_object_congestion_tracker.rs

```diff
@@ -11,117 +11,276 @@ use iota_types::{
+// transaction should be assigned. If true, the tracker will assign the start
+// time according to the minimum free execution slot for a transaction over all
+// its shared objects. If false, the tracker will assign the start time
+// according to the maximum end time of of the occupied execution slots for a
+// transaction over all its shared objects.
 #[derive(PartialEq, Eq, Clone, Debug)]
 pub struct SharedObjectCongestionTracker {
-    object_execution_cost: HashMap<ObjectID, u64>,
+    object_execution_slots: HashMap<ObjectID, Vec<ExecutionSlot>>,
     mode: PerObjectCongestionControlMode,
+    assign_min_free_execution_slot: bool,
 }
 
 impl SharedObjectCongestionTracker {
-    pub fn new(mode: PerObjectCongestionControlMode) -> Self {
+    pub fn new(mode: PerObjectCongestionControlMode, assign_min_free_execution_slot: bool) -> Self {
         Self {
-            object_execution_cost: HashMap::new(),
+            object_execution_slots: HashMap::new(),
             mode,
+            assign_min_free_execution_slot,
         }
     }
 
-    pub fn new_with_initial_value_for_test(
-        init_values: &[(ObjectID, u64)],
-        mode: PerObjectCongestionControlMode,
-    ) -> Self {
-        let mut object_execution_cost = HashMap::new();
-        for (object_id, total_cost) in init_values {
-            object_execution_cost.insert(*object_id, *total_cost);
+    // Given a list of shared input objects and the estimated execution duration of
+    // a transaction that operates on these objects, returns the starting time
+    // of the transaction
+    //
+    // Starting time is determined by all the input shared objects' last write.
+    pub fn compute_tx_start_time(
+        &mut self,
+        shared_input_objects: &[SharedInputObject],
+        tx_duration: ExecutionTime,
+    ) -> ExecutionTime {
+        // initialise the free execution slots for the objects that are not in the
+        // tracker.
+        for obj in shared_input_objects {
+            self.object_execution_slots
+                .entry(obj.id)
+                .or_insert(vec![ExecutionSlot::max_duration_non_scheduled_slot()]);
         }
-        Self {
-            object_execution_cost,
-            mode,
+        if self.assign_min_free_execution_slot {
+            // If `assign_min_free_execution_slot` is true, we assign the transaction start
+            // time based on the lowest free execution slot that can accommodates the
+            // transaction. We start the search from the full range of the slots
+            // available with no constraints from previous objects.
+            let initial_free_slot = ExecutionSlot::max_duration_non_scheduled_slot();
+            self.compute_min_free_execution_slot(
+                shared_input_objects,
+                tx_duration,
+                initial_free_slot,
+            )
+            .unwrap_or(MAX_EXECUTION_TIME)
+        } else {
```

Comment:
Perhaps we can remove this branch of code once the feature is enabled on the mainnet. Maybe we should mark such places with TODO?

Reply:
My understanding was that such branches activated by feature flags need to remain even after the flag is set to true on mainnet.

Reply:
Yes, so the old branch will never be used and we can remove it then I guess?

Reply:
But then we only have the new logic and the feature flag will also be unused anywhere. Are there any other cases where we do this?

Reply:
Take `native_charging_v2` flag for example. It has been enabled on mainnet since version 5 and presumably will remainenabled until the a v3 comes along, but we still keep the old logic in this case. https://github.com/iotaledger/iota/blob/943cbb213439dff8bb1b9761f65480f5729edc5e/iota-execution/latest/iota-move-natives/src/crypto/groth16.rs#L214

Reply:
It always comes down to the question: Do we need that protocol config parameter/flag to replay the transactions/execution in the exact same way or not? When it comes to consensus flags, those can be removed as soon as the feature got enabled, because the consensus is not replayed later on again. But the transaction execution should always be kept the same.